OpenSkyPlusApi/DeviceControls.cs

```diff
@@ -1,230 +1,380 @@
+                "MDOFGLDBEIK", BindingFlags.InvokeMethod | BindingFlags.Public | BindingFlags.Instance,
+                null, _launchMonitorWrapper, null);
+            Armed = true;
+            MessageMonitorReady?.Invoke();
+            return true;
+        }
+        catch
+        {
+            return false;
+        }
+    }
+
+    public static event NotificationMonitorStatusChange MessageMonitorNotReady;
+
+    public static bool DisarmMonitor()
+    {
+        _logger.LogInfo("DisarmMonitor");
+
+        // public void JDIMANKGGMF
+        try
+        {
+            _launchMonitorWrapper.GetType().InvokeMember(
+                "JDIMANKGGMF", BindingFlags.InvokeMethod | BindingFlags.Public | BindingFlags.Instance,
+                null, _launchMonitorWrapper, null);
+            Armed = false;
+            MessageMonitorNotReady?.Invoke();
+            return true;
+        }
+        catch
+        {
+            return false;
+        }
+    }
+
+    public static ShotMode SetShotMode(ShotMode shotMode)
+    {
+        _logger.LogInfo($"Setting shot mode to {shotMode}");
+
+        // Putting:
+        // public void CBJNBOIKFIE
+        //
+        // Normal:
+        // public void FPFBIPJNOKH
+        var wasArmed = Armed == true;
+        try
+        {
+            DisarmMonitor();
+            switch (shotMode)
+            {
+                case ShotMode.Putting:
+                    _launchMonitorWrapper.GetType().InvokeMember(
+                        "CBJNBOIKFIE", BindingFlags.InvokeMethod | BindingFlags.Public | BindingFlags.Instance,
+                        null, _launchMonitorWrapper, null);
+                    shotMode = ShotMode.Putting;
+                    return ShotMode.Putting;
+                case ShotMode.Normal:
+                default:
+                    _launchMonitorWrapper.GetType().InvokeMember(
+                        "FPFBIPJNOKH", BindingFlags.InvokeMethod | BindingFlags.Public | BindingFlags.Instance,
+                        null, _launchMonitorWrapper, null);
+                    return ShotMode.Normal;
+            }
+        }
+        catch (Exception ex)
+        {
+            _logger.LogWarning($"Couldn't set shot mode to {shotMode}:\n{ex}");
+            return shotMode;
+        }
+        finally
+        {
+            if (_config.AppSettings.RefreshConnectionAfterModeSwitch) RefreshConnection();
+            if (wasArmed) ArmMonitor();
+        }
+    }
+
+    public static async void RefreshConnection(bool disconnectOnly = false)
+    {
+        _logger.LogInfo("RefreshConnection: pause/unpause");
+
+        // Pause SW
+        _launchMonitorController.GetType()
+            .InvokeMember("JNCPBMDPIKB", BindingFlags.InvokeMethod | BindingFlags.Public | BindingFlags.Instance,
+                null, _launchMonitorController, [true]);
+
+        // Do not call twice if only want to pause (disconnect)
+        if (!disconnectOnly)
+        {
+            await Task.Delay(200);
```

Comment:
Does this prevent the reconnection failure?

Reply:
I think so, I figured this out by experiment so I am not sure of the optimal delay

Reply:
Cool. Might cause some weird state issues later if it's different for different connection types, but if it's working right now I see no problem to leave it as is.